lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -363,7 +400,55 @@ LookupTypeResult TypeChecker::lookupMemberType(DeclContext *dc,
   if (options.contains(NameLookupFlags::IgnoreAccessibility))
     subOptions |= NL_IgnoreAccessibility;
 
-  if (!dc->lookupQualified(type, name, subOptions, this, decls))
+  auto lookupSuccess = dc->lookupQualified(type, name, subOptions, this, decls);
+
+  // If the requested type is a CodingKeys type on something which may require
+  // Encodable/Decodable synthesis, it may not have been synthesized yet. If
+  // this is the case, force synthesis and attempt lookup again.
+  if (!lookupSuccess && name == Context.Id_CodingKeys) {
+    if (auto *targetDecl = type->getAnyNominal()) {
+      auto *encodableProto = Context.getProtocol(KnownProtocolKind::Encodable);
+      auto *decodableProto = Context.getProtocol(KnownProtocolKind::Decodable);
+
+      // If the type either conforms to Encodable/Decodable directly, or
+      // inherits from something which does, the conformance check will force
+      // synthesis, which will allow us to try again.
+      NormalProtocolConformance *conformance = nullptr;
+      Optional<ProtocolConformanceRef> conformanceRef;
+      if ((conformanceRef = conformsToProtocol(type, decodableProto, targetDecl,
+                                               ConformanceCheckFlags::Used,
+                                               SourceLoc()))) {
+        if (auto *concrete = conformanceRef->getConcrete()) {
+          conformance = dyn_cast<NormalProtocolConformance>(concrete);
+        }
+      }
+
+      if (!conformance &&
+          (conformanceRef = conformsToProtocol(type, encodableProto, targetDecl,
+                                               ConformanceCheckFlags::Used,
+                                               SourceLoc()))) {
+        if (auto *concrete = conformanceRef->getConcrete()) {
+          conformance = dyn_cast<NormalProtocolConformance>(concrete);
+        }
+      }
+
+      if (conformance) {
+        // Check conformance, forcing synthesis.
+        //
+        // If synthesizing conformance fails, this will produce diagnostics.
+        // However, these diagnostics are going to be produced again later when
+        // protocol conformance is being checked in a separate pass. To prevent
+        // this duplication, we'll throw the diagnostics away here.
```

Comment:
I'm not sure this will work as described. Once you've checked the conformance, it stays checked, and the diagnostics won't be emitted again, no? cc @DougGregor